### ISSUE TITLE
Duck DNS addon does not renew expired certificate

### ISSUE THREAD

**Issue by akmolina28** [home-assistant/addons#1955]
## The problem
I'm using the Duck DNS addon for SSL and my certificate has expired. The addon is not updating the certificate. The addon thinks that the certificate is still valid and does not renew it.

I have tried reinstalling the addon and deleting the local data to force it to renew, but the cert from Let's Encrypt still has the expired date.

## Environment

- Add-on with the issue: Duck DNS
- Add-on release with the issue: 1.12.5
- Last working add-on release (if known): 1.12.4
- Operating environment (OS/Supervised): OS

## Problem-relevant configuration

```yaml
lets_encrypt:
  accept_terms: true
  certfile: fullchain.pem
  keyfile: privkey.pem
token: [REDACTED FOR PRIVACY]
domains:
  - [REDACTED FOR PRIVACY].duckdns.org
aliases: []
seconds: 300
```

## Traceback/Error logs

Here is the output when I re-install and force it to generate everything.

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] done.
[services.d] starting services
[services.d] done.
# INFO: Using main config file /data/workdir/config
+ Generating account key...
+ Registering account key with ACME server...
+ Fetching account ID...
+ Done!
[17:59:19] INFO: OK
68.1.53.206
NOCHANGE
[17:59:19] INFO: Renew certificate for domains: ######.duckdns.org and aliases: 
# INFO: Using main config file /data/workdir/config
Processing ######.duckdns.org
 + Creating new directory /data/letsencrypt/######.duckdns.org ...
 + Signing domains...
 + Generating private key...
 + Generating signing request...
 + Requesting new certificate order from CA...
 + Received 1 authorizations URLs from the CA
 + Handling authorization for ######.duckdns.org
 + 1 pending challenge(s)
 + Deploying challenge tokens...
OK + Responding to challenge for ######.duckdns.org authorization...
 + Challenge is valid!
 + Cleaning challenge tokens...
OK + Requesting certificate...
 + Checking certificate...
 + Done!
 + Creating fullchain.pem...
 + Done!
[18:04:28] INFO: OK
68.1.53.206
NOCHANGE
[18:09:29] INFO: OK
68.1.53.206
NOCHANGE
[18:14:30] INFO: OK
68.1.53.206
NOCHANGE
```

Here is the output when I just restart the addon:

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] done.
[services.d] starting services
[services.d] done.
# INFO: Using main config file /data/workdir/config
+ Account already registered!
[18:18:25] INFO: OK
68.1.53.206
NOCHANGE
[18:18:26] INFO: Renew certificate for domains: ######.duckdns.org and aliases: 
# INFO: Using main config file /data/workdir/config
Processing ######.duckdns.org
 + Checking domain name(s) of existing cert... unchanged.
 + Checking expire date of existing cert...
 + Valid till Jun 30 21:59:27 2021 GMT Certificate will not expire
(Longer than 30 days). Skipping renew!
```

**Note that it says the certificate is valid till Jun 30. This is not true and I think this is the root of the problem.** I don't know where it is getting this Jun 30 date from, because the actual cert expired March 29 as shown here:

![image](https://user-images.githubusercontent.com/2997206/113363608-e8d88680-9316-11eb-9dba-3f409947958f.png)

## Additional information

I am also using the nginx proxy addon. Everything has been working fine for about 9 months. I did recently upgrade the addon before this issue started happening.

**Comment by akmolina28** [comment]
I think I fixed it. I think this had something to do with the Nginx addon -- I think it needed to be disabled to properly update the cert. Here is the procedure I used in case anyone else runs into this:

1. Disable DuckDNS AND Nginx addons

2. Add to configuration.yaml

```
http:
  base_url: https://yourdomain.duckdns.org
  ssl_certificate: /ssl/fullchain.pem
  ssl_key: /ssl/privkey.pem
```

3. Restart home assistant

4. Start DuckDNS -- it should fix the cert now

5. Remove the http node from the config again

6. Restart HA

7. Enable nginx again

At this point my setup was fixed and I was able to access locally over http and remotely over https as before. The cert from Let's Encrypt finally has a new expiry date.

**Comment by jeroenterheerdt** [comment]
thanks for explaining this - I had the same issue and was able to fix it using your steps, but had to remove/readd the DuckDNS addon before it would renew. also, I renamed the files in /ssl/ so I could see the renew happening.

**Comment by alstr** [comment]
Fixed my problem; I only needed to stop/start DuckDNS.

**Comment by mattmattmatt** [comment]
For myself googling this 3 months from now, and for those running into similar issues, this is what solved the certificate renewal for me:

1. Back up your DuckDNS Add-On Config somewhere
1. Remove these lines from the `http:` config:  
```
  ssl_certificate: /ssl/fullchain.pem
  ssl_key: /ssl/privkey.pem
```  
3. SSH into the host, remove or rename `/ssl/fullchain.pem` and `/ssl/privkey.pem`
1. Uninstall DuckDNS Add-On
1. Restart Home Assistant, access over `http`
1. Install DuckDNS Add-On
1. Paste backed up DuckDNS config, modify to have empty `aliases: []` key
1. Start DuckDNS Add-On, check logs to verify it successfully renewed the duckdns.org certificate
1. Paste backed up DuckDNS config with the original `aliases:` key content
1. Restart DuckDNS Add-On, check logs to verify it successfully renewed the certificate of the alias
1. Add the `ssl_certificate` and `ssl_key` lines from step 2 back into the `http:` config
1. Restart Home Assistant and access over `https`

It's entirely possible that not all of these steps are necessary to successfully renew the certificates, but these ones work for me.

**Comment by mattmattmatt** [comment]
Like clockwork, I'm back 3 months later to remind myself what to do after the automatic certificate renewal in the DuckDNS add-on failed with this response:
```JSON
{
  "type": "dns-01",
  "status": "invalid",
  "error": {
    "type": "urn:ietf:params:acme:error:unauthorized",
    "detail": "Incorrect TXT record \"xxxxxxxxx\" found at _acme-challenge.mmm.duckdns.org",
    "status": 403
  },
  "url": "https://acme-v02.api.letsencrypt.org/acme/chall-v3/mmm/mmm",
  "token": "mattmattmatt-mattmattmatt",
  "validated": "2022-07-27T08:40:31Z"
}
```

I verified that the following steps worked to let the add-on renew the certificates:
1. Back up DuckDNS Add-On config, stop the add-on, disable "start on boot"
2. Modify DuckDNS config to have empty `aliases: []` key
3. Remove/comment out these lines from the `http:` section in _configuration.yaml_:
```
  ssl_certificate: /ssl/fullchain.pem
  ssl_key: /ssl/privkey.pem
```
4. Restart Home Assistant, access over `http://`
5. Start DuckDNS Add-On, check its logs to verify it successfully renewed the duckdns.org certificate
6. Paste the original `aliases:` key content into the DuckDNS config, re-enable "start on boot"
7. Add back the `ssl_certificate` and `ssl_key` lines from step 2 into the `http:` config
8. Restart Home Assistant and access over `https://`

See you in October!

**Comment by ReX1983** [comment]
I just experienced the same issue: the certificate expired but Duck DNS log reported it as valid until some time in the future.

I tried to restart HA and Duck DNS but it didn't work, but then I rebooted the whole host and it worked.

My theory is that Duck DNS had actually renewed the certificate, but the older version was still used by HA and NGINX (maybe due to symlinks still pointing to the old key?).
